substrate/frame/staking/src/slashing.rs

```diff
@@ -336,31 +332,17 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
 
 		let validator_index_u32 = validator_index as u32;
 
-		match offending.binary_search_by_key(&validator_index_u32, |(index, _)| *index) {
+		if let Err(index) = offending.binary_search_by_key(&validator_index_u32, |index| *index) {
 			// this is a new offending validator
-			Err(index) => {
-				offending.insert(index, (validator_index_u32, disable));
 
-				let offending_threshold =
-					T::OffendingValidatorsThreshold::get() * validators.len() as u32;
-
-				if offending.len() >= offending_threshold as usize {
-					// force a new era, to select a new validator set
-					<Pallet<T>>::ensure_new_era()
-				}
+			// we don't want to disable too many validators otherwise we will break consensus
+			if offending.len() >= disabling_limit(validators.len()) as usize {
+				return
```

Comment:
For what its worth, I was not yet convinced that re-enabling causes any issues: At least on era changes, we get "re-enable" events anyway. I would also rather re-enable the lowest offenders once we reach the limit to not go over it, instead of just stopping disabling. Anyhow, it is an edge case that in theory does not even exist, so I will not die on this hill.

> Imagine the case where the byzantine threshold is filled in with 'small offenders' and at a later point a big offender needs to be disabled

This is precisely the case why we would want that. If an attacker found a consensus issue, it could fill up our threshold with honest disabled guys and then try to back malicious candidates. The attacker would now get slashed once, but could keep trying for an entire era at that flat rate -> NOT great. This would be impossible, if we made sure to _always_ have the highest offenders disabled.

> So we might have a case where node A considers an offender disabled while node B doesn't.

This is not a problem. Eventually nodes will agree. In the timespan they don't, it also does not pose any real issues as one goal of the new disabling strategy was to have disabling not affect consensus.

Reply:
In any case, I don't want to delay this feature. As not having this at all, is worse than any of the two options. I think what happens at the threshold should be an implementation detail, that can be changed later, if we want to. Considerations, should end up in the guide though.

Reply:
I like the idea of the swap disabled validator when the threshold is filled. The implementation doesn't seem too complicated, but If this is not critical for a 1st iteration, we could open an issue and work on it later. In any case, probably work checking with W3F people or so if not having the swap will considerably impact security (it doesn't seem to be the case though).

Reply:
I've created a follow up issue for this: https://github.com/paritytech/polkadot-sdk/issues/2418

Let's leave it as is for now.